### DISCUSSION TITLE
Configurable default log_level

### DISCUSSION THREAD

**Discussion by jurzik** [gabime/spdlog#3248] [Ideas]
I am using spdlog in a library, that is supposedly to be used by an external resource, partner or even my other application.
Once the actual application loads this library and tries to execute some of the API calls my library provides, the first call to logging function is invoked.
`spdlog::error(error_message);`

This by default logs info level and above into the console output.
Since I do not have any visibility on the application or environment, I would prefer if it would be possible to disable the logging level by default.
Technically, there are some static functions in the c++, that can be executed during the DLL is load into memory, however if there would be and error it could crash the entire application and I do not like the approach.

I have found the code, that is responsible for this in the file [registry.h](https://github.com/gabime/spdlog/blob/8e5613379f5140fefb0b60412fbf1f5406e7c7f8/include/spdlog/details/registry.h#L114)

`spdlog::level::level_enum global_log_level_ = level::info;`

Would it be possible to change this in a similar fashion to use compile time constant as you are doing for [SPDLOG_ACTIVE_LEVEL](https://github.com/gabime/spdlog/blob/8e5613379f5140fefb0b60412fbf1f5406e7c7f8/include/spdlog/common.h#L240)?

Something like
`#if !defined(SPDLOG_DEFAULT_LEVEL)`
`    #define SPDLOG_DEFAULT_LEVEL SPDLOG_LEVEL_INFO`
`#endif`

`spdlog::level::level_enum global_log_level_ = SPDLOG_DEFAULT_LEVEL;`

**Comment by tt4g** [comment]
`global_log_level_` can be changed with `spdlog::set_level()`.
There is no option to change it with a macro at compile time.

FYI: There are two ways to change `global_log_level_` at runtime: application arguments and environment variables.
 * `spdlog::cfg::load_argv_levels()`: https://github.com/gabime/spdlog/blob/v1.15.0/include/spdlog/cfg/argv.h
 * `spdlog::cfg::load_env_levels()`: https://github.com/gabime/spdlog/blob/v1.15.0/include/spdlog/cfg/env.h

**Reply by jurzik** [reply]
I am aware that this can be done at runtime, what I am arguing is the default value and the requirement on the final application.
3rd party application should not be enforced to disable, redirect or change this value by default - they should opt-in to enable the logging in my opinion.
We are going to spam console/syslog with info messages, that they are not aware are there.

Currently, there is no option in the compile time, however I am asking for such change as this is real scenario I am currently dealing with.
I am also willing to implement it, as this should be 1-hour work, but wanted to discuss beforehand.
Doing a change, that would be rejected does not make any sense.

**Reply by tt4g** [reply]
@gabime is the maintainer of this repository.
If @gabime does not respond, it may be faster to create a PR.

**Comment by gabime** [comment]
Whats wrong in setting this at runtime?

**Reply by jurzik** [reply]
I wouldn't call it wrong, however in my specific setup this creates unnecessary implication for consumer of my library.

To give broader context, I am developing a library to process PDF files with C interface,  that can be consumed by another solution.
Users of this library look at the API, call File_Open, do some work and File_Close within their application.
Within File_Open I log the name of file that is being opened with level info and it is immediately transferred to console output.
Consumers of this library might not be aware of it and in case they rely on console output to parse the data I will disrupt their results.

The only way I could tweak this configuration at runtime would be using DllMain on Windows, init and fini on Linux/MAC, while keeping in mind the differences between each compiler.
If there would be a standard main entry point for dynamic libraries I might consider this approach, but for the time being there is not.

I can write the information that the logging is enabled by default into the documentation and try to warn all the users about this.
If I have ever observed anything, it is that reading of documentation does never happen, so I would prefer to have it disabled by default.

In the initial idea I have identified the place where this could be changed using compile time constants in the same way you are using for [SPDLOG_ACTIVE_LEVEL](https://github.com/gabime/spdlog/blob/8e5613379f5140fefb0b60412fbf1f5406e7c7f8/include/spdlog/common.h#L240).
This change is backwards compatible as when the SPDLOG_DEFAULT_LEVEL is not defined, it would behave in the exact same manner.
